crates/cadenza-syntax/STATUS.md

```diff
@@ -0,0 +1,450 @@
+
+**Design Decisions Needed:**
+
+**Functions (hoisted):**
+- Named, can be called before definition
+- Syntax: `fn name params -> body`
+- Enables mutual recursion
+
+**Closures (not hoisted):**
+- Anonymous, bound to variables via `let`
+- Must be defined before use
+- Several syntax options:
+  - Bare arrow: `x -> x + 1`
+  - Backslash: `\x -> x + 1` (Haskell-like)
+  - Pipes: `|x| x + 1` (Rust-like)
+  - `fn` keyword: `fn x -> x + 1`
+
+**Open Questions:**
+1. Closure syntax choice
+2. Function name position (`fn double x` vs `double = fn x`)
+3. Parameter syntax (curried `fn add x y` vs tuple `fn add (x, y)`)
+
+**References:** `PARSER_ISSUES.md` Issue 10
+
+---
+
+### 11. Loops ❌ NOT IMPLEMENTED
+
+**Status:** No loop constructs.
+
+**What's Needed:**
+1. Parse `loop` with block body
+2. Parse `while` with condition and block
+3. Parse `for pattern in iterable` with block
+4. Add `break` and `continue` for loop control
+5. Optional: loop labels for nested loop control
+
+**Design Implications:** Supporting loops implies supporting mutability.
+
+**Syntax Examples:**
+```cadenza
+loop
+    do_something
+    if done then break
+
+while condition
+    do_something
+
+for x in collection
+    process x
+```
+
+**References:** `PARSER_ISSUES.md` Issue 11
+
+---
+
+### 12. Partial Function Application ❌ NOT IMPLEMENTED
+
+**Status:** No partial application support.
+
+**What's Needed:**
+1. Add `&` as prefix operator for partial application
+2. Parse function name and following arguments
+3. Handle `$0`, `$1`, `$2` as positional placeholders
+4. Represent as `Apply(&, [fn, arg1, arg2, ...])`
+5. Handle holes for argument reordering
+
+**Design Conflict:** `&` also used for bitwise AND (infix) and potential reference operator.
+
+**Syntax Examples:**
+```cadenza
+let f = &add          # Capture function
+let add5 = &add 5     # Partial application
+let middle = &substring 0 $0 10  # With holes
+```
+
+**References:** `PARSER_ISSUES.md` Issue 12
+
+---
+
+### 13. Error Recovery ⚠️ PARTIAL
+
+**Status:** Basic error nodes exist, recovery is weak.
+
+**Issues:**
+1. `parse_primary()` has catch-all that accepts invalid tokens
+2. Missing synchronization strategy after errors
+3. Limited negative test coverage
+4. Errors in delimited contexts don't propagate correctly
+
+**What's Needed:**
+1. Restrict valid primary tokens (identifiers, literals, delimiters, prefix ops)
+2. Create Error nodes for unexpected tokens
+3. Implement synchronization (skip to newline, delimiter, etc.)
+4. Add comprehensive negative tests
+5. Fix marker propagation to handle dedented expressions in delimiters
+
+**Example Issue:** `foo [\nbar` where `bar` is at root indentation should not become an argument of `foo`.
+
+**References:** `PARSER_ISSUES.md` Issue 13
+
+---
+
+### 14. String Interpolation ❌ NOT IMPLEMENTED
+
+**Status:** Basic strings work, no interpolation or heredoc support.
+
+**What's Needed:**
+1. Add `:` prefix for interpolated strings: `:"hello ${name}"`
```

Comment:
I'm actually wondering if we use the same thing as JS. So `hello ${name}`. I'm wanting to reserve `:` for type annotations on `let v: integer = 1`